pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -716,20 +713,39 @@ func (q *queue) processPartition(ctx context.Context, p *QueuePartition, shard *
 	// within 5ms of each other, we fetch them in order but we may process them out of
 	// order, depending on how long it takes for the item to pass through the channel
 	// to the worker, how long Redis takes to lease the item, etc.
-	fetch := time.Now().Truncate(time.Second).Add(PartitionLookahead)
+	fetch := getNow().Truncate(time.Second).Add(PartitionLookahead)
 	queue, err := q.Peek(peekCtx, p.Queue(), fetch, q.peekSize())
 	if err != nil {
 		return err
 	}
 
-	var processErr error
-	var concurrencyLimitReached bool
+	var (
+		processErr error
+
+		// These flags are used to handle partition rqeueueing.
+		ctrSuccess     int32
+		ctrConcurrency int32
+		ctrRateLimit   int32
```

Comment:
nit: non-blocker

probably good to wrap these variables into a struct if it's gonna get more complicated than this. easier to create helper functions as well instead of doing things like
```
if ctrConcurrency > 0 || (ctrRateLimit > 0 && ctrConcurrency == 0 && ctrSuccess == 0)
```
which is not intuitive on first sight, and require comments to clarify intent.